### ISSUE TITLE
Debian package upgrade failing - unsupported action: upgrade

### ISSUE THREAD

**Issue by kwisatz** [grafana/mimir#6013]
#### Describe the bug
Upgrading the mimir package on debian systems is failing.

#### To Reproduce

1. Have packages.grafana.com as source for apt
2. apt update
3. apt upgrade

#### Expected behavior

Upgrade should not fail

#### Environment

```bash
(new)root@axanar:~# cat /etc/debian_version
11.7
(new)root@axanar:~# apt-cache policy mimir
mimir:
  Installed: 2.7.2
  Candidate: 2.9.0
```

#### Additional Context

```bash
(new)root@axanar:~# apt upgrade
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
The following packages will be upgraded:
  mimir
1 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/21.1 MB of archives.
After this operation, 6,399 kB of additional disk space will be used.
Do you want to continue? [Y/n]
Reading changelogs... Done
(Reading database ... 68271 files and directories currently installed.)
Preparing to unpack .../archives/mimir_2.9.0_amd64.deb ...
Unsupported action: upgrade
dpkg: warning: old mimir package pre-removal script subprocess returned error exit status 1
dpkg: trying script from the new package instead ...
Unsupported action: failed-upgrade
dpkg: error processing archive /var/cache/apt/archives/mimir_2.9.0_amd64.deb (--unpack):
 new mimir package pre-removal script subprocess returned error exit status 1
 Alpine Post Install of an clean install
 Reload the service unit from disk
 Unmask the service
 Set the preset flag for the service unit
 Set the enabled flag for the service unit
Errors were encountered while processing:
 /var/cache/apt/archives/mimir_2.9.0_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)

```

**Comment by 56quarters** [comment]
Sounds very similar to https://github.com/grafana/mimir/pull/4801

**Comment by aknuds1** [comment]
@wilfriedroset could you have a look?

**Comment by wilfriedroset** [comment]
@56quarters is right. The packaging was broken prior 2.8, the fix is right here https://github.com/grafana/mimir/pull/4801/files#diff-fd303934e8f1839d3ac71bc5475dec3f641c284f54089f1c6af563003d873f75R16-R19
Unfortunately there is no other solution apart from `apt remove mimir && apt install mimir`

Here is a reproducer and an example of how to get out of this situation.
```
root@test:~# apt-get install -y apt-transport-https software-properties-common wget gnupg
root@test:~# mkdir -p /etc/apt/keyrings/
root@test:~# wget -q -O - https://apt.grafana.com/gpg.key | gpg --dearmor | sudo tee /etc/apt/keyrings/grafana.gpg > /dev/null
root@test:~# echo "deb [signed-by=/etc/apt/keyrings/grafana.gpg] https://apt.grafana.com stable main" | sudo tee -a /etc/apt/sources.list.d/grafana.list
root@test:~# apt-get update
```
```
root@test:~# apt-get install mimir=2.7.2
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  mimir
0 upgraded, 1 newly installed, 0 to remove and 4 not upgraded.
Need to get 19.5 MB of archives.
After this operation, 53.8 MB of additional disk space will be used.
Get:1 https://apt.grafana.com stable/main amd64 mimir amd64 2.7.2 [19.5 MB]
Fetched 19.5 MB in 3s (6794 kB/s)
Selecting previously unselected package mimir.
(Reading database ... 21984 files and directories currently installed.)
Preparing to unpack .../archives/mimir_2.7.2_amd64.deb ...
Unpacking mimir (2.7.2) ...
Setting up mimir (2.7.2) ...
 Post Install of an clean install
 Reload the service unit from disk
 Unmask the service
 Set the preset flag for the service unit
Created symlink /etc/systemd/system/multi-user.target.wants/mimir.service → /lib/systemd/system/mimir.service.
 Set the enabled flag for the service unit

root@test:~# mimir --version
Mimir, version 2.7.2 (branch: release-2.7, revision: 76a1021a1)
  go version:       go1.20.3
  platform:         linux/amd64

root@test:~# apt-cache policy mimir
mimir:
  Installed: 2.7.2
  Candidate: 2.9.0
  Version table:
     2.9.0 500
        500 https://apt.grafana.com stable/main amd64 Packages
     2.8.0 500
        500 https://apt.grafana.com stable/main amd64 Packages
     2.8.0~rc.0 500
        500 https://apt.grafana.com stable/main amd64 Packages
     2.7.3 500
        500 https://apt.grafana.com stable/main amd64 Packages
 *** 2.7.2 500
        500 https://apt.grafana.com stable/main amd64 Packages
        100 /var/lib/dpkg/status
     2.7.1 500
        500 https://apt.grafana.com stable/main amd64 Packages
     2.6.2 500
        500 https://apt.grafana.com stable/main amd64 Packages
     2.6.1 500
        500 https://apt.grafana.com stable/main amd64 Packages
     2.6.0 500
        500 https://apt.grafana.com stable/main amd64 Packages
     2.5.0 500
        500 https://apt.grafana.com stable/main amd64 Packages
     2.4.0 500
        500 https://apt.grafana.com stable/main amd64 Packages
     2.3.1 500
        500 https://apt.grafana.com stable/main amd64 Packages
     2.3.0 500
        500 https://apt.grafana.com stable/main amd64 Packages
```
At this stage we have a similar setup than the one reported by @kwisatz 
```
root@test:~# apt upgrade
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
The following packages will be upgraded:
  mimir
1 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 21.1 MB of archives.
After this operation, 6399 kB of additional disk space will be used.
Do you want to continue? [Y/n] y
Get:1 https://apt.grafana.com stable/main amd64 mimir amd64 2.9.0 [21.1 MB]
Fetched 21.1 MB in 3s (6925 kB/s)
Reading changelogs... Done
(Reading database ... 23181 files and directories currently installed.)
Preparing to unpack .../archives/mimir_2.9.0_amd64.deb ...
Unsupported action: upgrade
dpkg: warning: old mimir package pre-removal script subprocess returned error exit status 1
dpkg: trying script from the new package instead ...
Unsupported action: failed-upgrade
dpkg: error processing archive /var/cache/apt/archives/mimir_2.9.0_amd64.deb (--unpack):
 new mimir package pre-removal script subprocess returned error exit status 1
 Alpine Post Install of an clean install
 Reload the service unit from disk
 Unmask the service
 Set the preset flag for the service unit
 Set the enabled flag for the service unit
Errors were encountered while processing:
 /var/cache/apt/archives/mimir_2.9.0_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)
```
We can confirm that we cannot upgrade because the preremove script is not correct in the 2.7.2 package. We can only remove the package (at this point you might want to backup your configuration just in case)
```
root@test:~# apt remove mimir
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following packages will be REMOVED:
  mimir
0 upgraded, 0 newly installed, 1 to remove and 0 not upgraded.
After this operation, 53.8 MB disk space will be freed.
Do you want to continue? [Y/n] y
(Reading database ... 23180 files and directories currently installed.)
Removing mimir (2.7.2) ...
root@test:~# apt purge mimir
--8<--
```
The fix is supposed to be in 2.8+, let's try it
```
root@test:~# apt install mimir=2.8.0
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  mimir
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 21.0 MB of archives.
After this operation, 59.7 MB of additional disk space will be used.
Get:1 https://apt.grafana.com stable/main amd64 mimir amd64 2.8.0 [21.0 MB]
Fetched 21.0 MB in 2s (11.4 MB/s)
Selecting previously unselected package mimir.
(Reading database ... 23174 files and directories currently installed.)
Preparing to unpack .../archives/mimir_2.8.0_amd64.deb ...
Unpacking mimir (2.8.0) ...
Setting up mimir (2.8.0) ...
 Post Install of an clean install
 Reload the service unit from disk
 Unmask the service
 Set the preset flag for the service unit
 Set the enabled flag for the service unit
```
Now that we have installed the 2.8.0 let's try to upgrade to the latest, in this case 2.9.0
```
root@test:~# apt upgrade
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
The following packages will be upgraded:
  mimir
1 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/21.1 MB of archives.
After this operation, 480 kB of additional disk space will be used.
Do you want to continue? [Y/n] y
Reading changelogs... Done
(Reading database ... 23181 files and directories currently installed.)
Preparing to unpack .../archives/mimir_2.9.0_amd64.deb ...
Unpacking mimir (2.9.0) over (2.8.0) ...
Setting up mimir (2.9.0) ...
 Post Install of an upgrade
 Post Install of an upgrade
Warning: The unit file, source configuration file or drop-ins of mimir.service changed on disk. Run 'systemctl daemon-reload' to reload units.
root@test:~# mimir --version
Mimir, version 2.9.0 (branch: release-2.9, revision: 761114d8b)
  go version:       go1.20.5
  platform:         linux/amd64
```

@kwisatz let me know if you this works for you. In my end I've managed to rollout the upgrade with ansible.

**Comment by kwisatz** [comment]
Yes, @wilfriedroset I can confirm that removing, then installing anew works for me.
Thanks
